packages/exp/src/commands/upload/BaseUploader.js

```diff
@@ -0,0 +1,224 @@
+    ];
+    const { source } = await prompt({
+      name: 'source',
+      message: 'Which build do you want to upload?',
+      type: 'list',
+      choices: SOURCES_LIST,
+    });
+
+    if (source === 'latest') {
+      return this.latestGetter();
+    } else {
+      return this.chooseBuild();
+    }
+  }
+
+  async getData() {
+    const options = this.options;
+    if (options.path) {
+      return this.pathGetter();
+    } else if (options.latest) {
+      return this.latestGetter();
+    } else if (options.id) {
+      return this.idGetter();
+    } else {
+      return this.askUser();
+    }
+  }
+
+  async getConfigData() {
+    const { exp } = await ProjectUtils.readConfigJsonAsync(this.projectDir);
+    const configName = await ProjectUtils.configFilenameAsync(this.projectDir);
+    if (!exp) {
+      throw new Error(`Couldn't read ${configName} file in project at ${this.projectDir}.`);
+    }
+    this.ensureConfigDataIsCorrect(exp);
+    return exp;
+  }
+
+  async _downloadFile(src, dest) {
+    const response = await Axios({
+      method: 'GET',
+      url: src,
+      responseType: 'stream',
+    });
+    const totalLength = parseInt(response.headers['content-length'], 10);
+    const bar = new ProgressBar('Downloading [:bar] :percent :etas', {
+      complete: '=',
+      incomplete: ' ',
+      total: totalLength,
+    });
+    response.data.pipe(fs.createWriteStream(dest));
+    return new Promise((resolve, reject) => {
+      response.data.on('end', () => resolve(dest));
+      response.data.on('data', data => bar.tick(data.length));
+      response.data.on('error', error => reject(new Error(`${error}`)));
+    });
+  }
+
+  async downloadFile(remoteUrl) {
+    const dest = _.last(remoteUrl.split('/'));
+    if (fs.existsSync(dest)) {
+      log.warn(
+        `File ${dest} exists. If it's not ${this
+          .platformExtension} you want to upload, please change its name.`
+      );
+      return dest;
+    }
+    log(`Downloading build from ${remoteUrl}`);
+    return this._downloadFile(remoteUrl, dest);
+  }
+
+  async upload() {
+    this.ensureOptionsAreCorrect();
+    const data = await this.getData();
+    const configData = await this.getConfigData();
+    const path = data.path ? data.path : await this.downloadFile(data.remoteUrl);
+    const platformData = await this.getPlatformData();
+    await this.uploadToStore(configData, platformData, path);
+  }
+
+  ensureOptionsAreCorrect() {
+    const options = this.options;
+    const definedKeys = Object.keys(_.pick(options, OPTIONS)).filter(key => !_.isNil(options[key]));
+    if (definedKeys.length > 1) {
+      throw new Error(`You have to choose only one of --path, --id, --latest`);
+    }
+    if (options.path && !fs.existsSync(options.path)) {
+      throw new Error(`File ${options.path} doesn't exist`);
+    }
+    if (options.path) {
+      this.ensureFileIsCorrect(options.path);
+    }
+    this.ensurePlatformOptionsAreCorrect();
+  }
+
+  getFastlane() {
```

Comment:
Please don't do that, getter functions are not a good thing in JS. You can move this require to the constructor.